src/tracing/on_disk.rs

```diff
@@ -0,0 +1,98 @@
+use std::collections::HashSet;
+use std::sync::Mutex;
+use std::{fs, io};
+
+use anyhow::Context as _;
+use serde::{Deserialize, Serialize};
+use tracing_subscriber::prelude::*;
+use tracing_subscriber::{filter, fmt, registry};
+
+use super::*;
+
+#[derive(Clone, Debug, Default, Eq, PartialEq, Deserialize, Serialize)]
+#[serde(default)]
+pub struct Config {
+    pub enabled: Option<bool>,
+    pub log_file: Option<String>,
+    pub log_level: Option<String>,
+    pub span_events: Option<HashSet<config::SpanEvent>>,
+}
+
+impl Config {
+    pub fn merge(&mut self, other: Self) {
+        self.enabled = other.enabled.or(self.enabled.take());
+        self.log_file = other.log_file.or(self.log_file.take());
+        self.log_level = other.log_level.or(self.log_level.take());
+        self.span_events = other.span_events.or(self.span_events.take());
+    }
+}
+
+#[rustfmt::skip] // `rustfmt` formats this into unreadable single line :/
+pub type Logger<S> = filter::Filtered<
+    Option<Layer<S>>,
+    filter::EnvFilter,
+    S,
+>;
+
+#[rustfmt::skip] // `rustfmt` formats this into unreadable single line :/
+pub type Layer<S> = fmt::Layer<
+    S,
+    fmt::format::DefaultFields,
+    fmt::format::Format,
+    MakeWriter,
+>;
+
+pub type MakeWriter = Mutex<io::BufWriter<fs::File>>;
+
+pub fn new_logger<S>(config: &mut Config) -> Logger<S>
+where
+    S: tracing::Subscriber + for<'span> registry::LookupSpan<'span>,
+{
+    let layer = match new_layer(config) {
+        Ok(layer) => layer,
+        Err(err) => {
+            eprintln!(
+                "failed to enable logging into {} log-file: {err}",
+                config.log_file.as_deref().unwrap_or(""),
+            );
+
+            config.enabled = Some(false);
```

Comment:
for the record, I do have doubts about this one. I understand the reasoning why we don't want to crash service because of config and I do understand why we want to have this config in disabled state afterwards. But the whole pipeline of mutating config seems unpleasant